src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/RefDictionary.cs

```diff
@@ -0,0 +1,147 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Collections.Extensions
+{
+    public sealed class RefDictionary<TKey, TValue> : IEnumerable<(TKey Key, TValue Value)> where TValue : new()
+    {
+        private int _count;
+        private int[] _buckets;
+        private Entry[] _entries;
+        private IEqualityComparer<TKey> _comparer;
+
+        private struct Entry
+        {
+            public int next;
+            public TKey key;
+            public TValue value;
+        }
+
+        public RefDictionary(int capacity, IEqualityComparer<TKey> comparer)
+        {
+            _buckets = new int[capacity];
+            _entries = new Entry[capacity];
+            _comparer = comparer ?? EqualityComparer<TKey>.Default;
+        }
+
+        public RefDictionary(IEqualityComparer<TKey> comparer) : this(16, comparer) { }
+        public RefDictionary(int capacity) : this(capacity, null) { }
+        public RefDictionary() : this(null) { }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetBucketIndex(TKey key) => key.GetHashCode() & 0x7FFFFFFF % _buckets.Length;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetEntryIndex(TKey key) => _buckets[GetBucketIndex(key)] - 1;
+
+        public bool TryGetValue(TKey key, out TValue value)
```

Comment:
Can this return a `ref out` value parameter?

Reply:
An entry would have to be created. This method is just to get the value or default without create.

I'm not sure about a speclet. I'm getting close to finishing some perf ideas and need to add another GroupBy/AggregateBy/CountBy perf test example. Maybe if I present the results and we can discuss next steps then.